src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,584 @@
+     * performing the derive is not selected until the method is called. Once a
+     * provider is selected, it cannot be changed.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object (may not be
+     *     {@code null})
+     * @param kdfParameterSpec
+     *     derivation parameters (may not be {@code null}
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the
+     *     KDF output and according to the derivation parameters
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived
+     * @throws NullPointerException
+     *     if {@code alg} or {@code kdfParameterSpec} is null
+     */
+    public SecretKey deriveKey(String alg, KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+        synchronized (lock) {
+            if (alg == null || alg.isEmpty()) {
+                throw new NullPointerException(
+                    "the algorithm for the SecretKey return value may not be "
+                    + "null or empty");
+            }
+            Objects.requireNonNull(kdfParameterSpec);
+            if (spi != null) {
+                return spi.engineDeriveKey(alg, kdfParameterSpec);
+            }
+            Exception lastException = null;
+            while ((firstService != null) || serviceIterator.hasNext()) {
+                Service s;
+                if (firstService != null) {
+                    s = firstService;
+                    firstService = null;
+                } else {
+                    s = serviceIterator.next();
+                }
+                if (!JceSecurity.canUseProvider(s.getProvider())) {
+                    continue;
+                }
+                try {
+                    KDFSpi spi = (KDFSpi) s.newInstance(algorithmParameterSpec);
+                    SecretKey result = spi.engineDeriveKey(alg,
+                                                           kdfParameterSpec);
+                    provider = s.getProvider();
+                    this.spi = spi;
+                    firstService = null;
+                    serviceIterator = null;
+                    return result;
+                } catch (Exception e) {
+                    if (lastException == null) {
+                        lastException = e;
+                    }
+                }
+            }
+            // no working provider found, fail
+            if (lastException instanceof InvalidParameterSpecException) {
+                throw (InvalidParameterSpecException) lastException;
+            }
+            if (lastException instanceof RuntimeException) {
+                throw (RuntimeException) lastException;
+            }
+        }
+        // should never reach here
+        return null;
```

Comment:
I think this could reach here if `lastException` is `null`. In any case, it's safer to throw an exception here instead of returning null as this is clearly an error condition. Something like:

```
            throw new InvalidParameterSpecException
                ("No installed provider supports the deriveKey method with these parameters");
```

Same comment for `deriveData`.